File: src/spiral/mcp.py
from __future__ import annotations

import json
import os
import subprocess
from typing import Any


class MCPClient:
    def __init__(self, server_config: dict[str, Any]):
        self.name = server_config.get("name", "unknown")
        self.transport = server_config.get("transport", "stdio")
        self.command = server_config.get("command")
        self.args: list[str] = server_config.get("args", [])
        self.env: dict[str, str] = server_config.get("env", {})
        self.url = server_config.get("url")
        self._process: subprocess.Popen | None = None
        self._tools: list[dict[str, Any]] = []

    def connect(self) -> bool:
        if self.transport == "stdio" and self.command:
            full_env = {**os.environ, **self.env}
            self._process = subprocess.Popen(
                [self.command, *self.args],
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
                env=full_env,
            )
            return self._handshake()
        return False

    def _handshake(self) -> bool:
        init_msg = {
            "jsonrpc": "2.0",
            "id": 0,
            "method": "initialize",
            "params": {
                "protocolVersion": "2024-11-05",
                "capabilities": {},
                "clientInfo": {"name": "spiral", "version": "0.1.0"},
            },
        }
        response = self._send(init_msg)
        if response and response.get("result"):
            list_msg = {
                "jsonrpc": "2.0",
                "id": 1,
                "method": "tools/list",
                "params": {},
            }
            tools_response = self._send(list_msg)
            if tools_response and tools_response.get("result"):
                self._tools = tools_response["result"].get("tools", [])
                return True
        return False

    def _send(self, msg: dict[str, Any]) -> dict[str, Any] | None:
        if not self._process or not self._process.stdin:
            return None
        try:
            self._process.stdin.write(json.dumps(msg) + "\n")
            self._process.stdin.flush()
            line = self._process.stdout.readline()
            if line:
                return json.loads(line)
        except (OSError, json.JSONDecodeError):
            return None
        return None

    def list_tools(self) -> list[dict[str, Any]]:
        return self._tools

    def call_tool(self, name: str, args: dict[str, Any]) -> str:
        msg = {
            "jsonrpc": "2.0",
            "id": 2,
            "method": "tools/call",
            "params": {"name": name, "arguments": args},
        }
        response = self._send(msg)
        if response and response.get("result"):
            content = response["result"].get("content", [])
            texts = [
                block.get("text", "")
                for block in content
                if block.get("type") == "text"
            ]
            return "\n".join(texts)
        return f"Error: tool call failed for {name}"

    def disconnect(self) -> None:
        if self._process:
            self._process.terminate()
            self._process = None


class MCPManager:
    def __init__(self, config: Any):
        self.config = config
        self.clients: list[MCPClient] = []
        servers_json = os.getenv("SPIRAL_MCP_SERVERS", "[]")
        try:
            servers = json.loads(servers_json)
        except json.JSONDecodeError:
            servers = []
        for s in servers:
            client = MCPClient(s)
            self.clients.append(client)

    def connect_all(self) -> list[str]:
        connected: list[str] = []
        for c in self.clients:
            if c.connect():
                connected.append(c.name)
        return connected

    def get_all_tools(self) -> list[dict[str, Any]]:
        tools: list[dict[str, Any]] = []
        for c in self.clients:
            for t in c.list_tools():
                tools.append(
                    {
                        "type": "function",
                        "function": {
                            "name": f"mcp_{c.name}_{t['name']}",
                            "description": t.get("description", ""),
                            "parameters": t.get(
                                "inputSchema", {"type": "object", "properties": {}}
                            ),
                        },
                    }
                )
        return tools

    def call_tool(self, full_name: str, args: dict[str, Any]) -> str:
        for c in self.clients:
            prefix = f"mcp_{c.name}_"
            if full_name.startswith(prefix):
                tool_name = full_name[len(prefix) :]
                return c.call_tool(tool_name, args)
        return f"Error: MCP tool '{full_name}' not found"

    def disconnect_all(self) -> None:
        for c in self.clients:
            c.disconnect()

File: src/managers/mcp.ts
import { spawn, ChildProcess } from "node:child_process";

export interface MCPServerConfig {
  name: string;
  transport: "stdio" | "http";
  command?: string;
  args?: string[];
  url?: string;
  headers?: Record<string, string>;
  env?: Record<string, string>;
}

interface MCPToolInfo {
  server: string;
  name: string;
  description: string;
  parameters: Record<string, unknown>;
}

interface MCPConnection {
  process: ChildProcess;
  nextId: number;
  pending: Map<number, { resolve: (value: unknown) => void; reject: (reason: Error) => void }>;
  buffer: string;
  initialized: boolean;
}

export class MCPManager {
  private servers: Map<string, MCPServerConfig> = new Map();
  private connections: Map<string, MCPConnection> = new Map();
  private tools: Map<string, MCPToolInfo> = new Map();

  addServer(config: MCPServerConfig): void {
    this.servers.set(config.name, config);
  }

  removeServer(name: string): void {
    this.servers.delete(name);
    // If connected, disconnect first
    const conn = this.connections.get(name);
    if (conn) {
      conn.process.kill();
      this.connections.delete(name);
    }
    // Remove tools belonging to this server
    for (const [toolKey, info] of this.tools) {
      if (info.server === name) {
        this.tools.delete(toolKey);
      }
    }
  }

  listServers(): MCPServerConfig[] {
    return Array.from(this.servers.values());
  }

  listTools(): MCPToolInfo[] {
    return Array.from(this.tools.values());
  }

  async connect(name: string): Promise<boolean> {
    const config = this.servers.get(name);
    if (!config) {
      throw new Error(`MCP server '${name}' not configured`);
    }
    if (this.connections.has(name)) {
      return true; // Already connected
    }
    if (config.transport === "stdio") {
      return this.connectStdio(name, config);
    }
    // HTTP transport not implemented yet
    throw new Error(`HTTP transport not yet supported for MCP server '${name}'`);
  }

  private connectStdio(name: string, config: MCPServerConfig): Promise<boolean> {
    if (!config.command) {
      throw new Error(`stdio transport requires 'command' for server '${name}'`);
    }
    const child = spawn(config.command, config.args ?? [], {
      stdio: ["pipe", "pipe", "pipe"],
      env: { ...process.env, ...(config.env ?? {}) },
    });

    const conn: MCPConnection = {
      process: child,
      nextId: 1,
      pending: new Map(),
      buffer: "",
      initialized: false,
    };

    this.connections.set(name, conn);

    child.stdout?.on("data", (data: Buffer) => {
      conn.buffer += data.toString();
      this.processMessages(name, conn);
    });

    child.stderr?.on("data", (data: Buffer) => {
      // Log stderr but don't fail
      console.error(`[MCP ${name}] ${data.toString().trim()}`);
    });

    child.on("error", (err: Error) => {
      console.error(`[MCP ${name}] process error: ${err.message}`);
      for (const [, { reject }] of conn.pending) {
        reject(new Error(`MCP server '${name}' process error: ${err.message}`));
      }
      conn.pending.clear();
      this.connections.delete(name);
    });

    child.on("exit", (code: number | null) => {
      console.error(`[MCP ${name}] process exited with code ${code}`);
      for (const [, { reject }] of conn.pending) {
        reject(new Error(`MCP server '${name}' exited`));
      }
      conn.pending.clear();
      this.connections.delete(name);
    });

    // Send initialize request
    return this.sendRequest(name, "initialize", {
      protocolVersion: "2024-11-05",
      capabilities: {},
      clientInfo: { name: "spiral", version: "1.0.0" },
    })
      .then(async () => {
        conn.initialized = true;
        // After initialize, send initialized notification
        this.sendNotification(name, "notifications/initialized", {});
        // List tools
        await this.refreshTools(name);
        return true;
      })
      .catch((err: Error) => {
        this.connections.delete(name);
        child.kill();
        throw err;
      });
  }

  private async refreshTools(name: string): Promise<void> {
    const result = (await this.sendRequest(name, "tools/list", {})) as {
      tools?: Array<{
        name: string;
        description?: string;
        inputSchema?: Record<string, unknown>;
      }>;
    };
    const tools = result.tools ?? [];
    for (const tool of tools) {
      const toolKey = `${name}:${tool.name}`;
      this.tools.set(toolKey, {
        server: name,
        name: tool.name,
        description: tool.description ?? "",
        parameters: tool.inputSchema ?? {},
      });
    }
  }

  async disconnect(name: string): Promise<void> {
    const conn = this.connections.get(name);
    if (!conn) return;
    conn.process.kill();
    this.connections.delete(name);
    // Remove tools for this server
    for (const [toolKey, info] of this.tools) {
      if (info.server === name) {
        this.tools.delete(toolKey);
      }
    }
  }

  async disconnectAll(): Promise<void> {
    for (const name of Array.from(this.connections.keys())) {
      await this.disconnect(name);
    }
  }

  async callTool(server: string, toolName: string, args: Record<string, unknown>): Promise<string> {
    const conn = this.connections.get(server);
    if (!conn) {
      throw new Error(`MCP server '${server}' not connected`);
    }
    const result = (await this.sendRequest(server, "tools/call", {
      name: toolName,
      arguments: args,
    })) as { content?: Array<{ type: string; text?: string }> } | string;
    // Extract text from result
    if (typeof result === "string") return result;
    if (result?.content && Array.isArray(result.content)) {
      return result.content.map((c: { type: string; text?: string }) => c.text ?? "").join("\n");
    }
    return JSON.stringify(result);
  }

  private sendRequest(
    server: string,
    method: string,
    params: Record<string, unknown>,
  ): Promise<unknown> {
    const conn = this.connections.get(server);
    if (!conn) {
      return Promise.reject(new Error(`MCP server '${server}' not connected`));
    }
    const id = conn.nextId++;
    const msg = JSON.stringify({ jsonrpc: "2.0", id, method, params });
    return new Promise((resolve, reject) => {
      conn.pending.set(id, { resolve, reject: reject as (reason: Error) => void });
      conn.process.stdin?.write(msg + "\n");
    });
  }

  private sendNotification(server: string, method: string, params: Record<string, unknown>): void {
    const conn = this.connections.get(server);
    if (!conn) return;
    const msg = JSON.stringify({ jsonrpc: "2.0", method, params });
    conn.process.stdin?.write(msg + "\n");
  }

  private processMessages(_server: string, conn: MCPConnection): void {
    // Split on newlines — each line is a JSON-RPC message
    while (true) {
      const idx = conn.buffer.indexOf("\n");
      if (idx === -1) break;
      const line = conn.buffer.slice(0, idx).trim();
      conn.buffer = conn.buffer.slice(idx + 1);
      if (!line) continue;
      try {
        const msg = JSON.parse(line) as {
          id?: number;
          method?: string;
          result?: unknown;
          error?: { message: string };
        };
        if (msg.id !== undefined && (msg.result !== undefined || msg.error !== undefined)) {
          const pending = conn.pending.get(msg.id);
          if (pending) {
            conn.pending.delete(msg.id);
            if (msg.error) {
              pending.reject(new Error(msg.error.message));
            } else {
              pending.resolve(msg.result);
            }
          }
        }
      } catch {
        // Not valid JSON, skip
      }
    }
  }
}
